README.md
# IXP-Watch

IXP-watch script - is a tool for IXPs to continuously monitor layer 2 traffic on the exchange.

As well as storing a regular traffic sample, it will generate alerts for the following:-

- Excessive ARP
- Excessive traffic captured
- Spanning Tree
- Non-IP/IPv6 Traffic (for example CDP)
- Multicast/Traffic directed to 255.255.255.255 - DHCP/OSPF/IGP etc.
- Stray SNMP

The report contains stats on:

- ARP Queries / ARPs per min
- ICMP/ICMPv6 Packets
- ARPs Sponged
- Dead BGP Peers (bgp open attempts to sponged addresses)
- IPv6: Multicast Listener Query / Router Advertisement / Neighbor Advertisement
- IPv6: Multicast Listener Discovery (MLDv2)

- Top 30 Arpers
- Top 30 ARP destinations requested
- ARP sponge reply activity
- Non-Unicast IP traffic
- Non-IP Traffic
- ICMP messages

Optionally, metrics can be fed to a monitoring system or grapher.
(Script includes some examples for rrd.)

If you are using IXP Manager, ARP sponges can be automated instead of using the manual
sponge tool included. See the directory ./IXP-Manager for example scripts.

# Installation

- See the file INSTALL.TXT for more installation details.
- INTRO.TXT provides more background info on the original ixpwatch tool

# Install Script

To make installation easier, there is a new (beta) installation script.
This has been tested on Debian/Ubuntu, but may work on other Linux distros.

wget -O install.sh https://raw.githubusercontent.com/euro-ix/IXP-Watch/master/install/install.sh && sudo bash ./install.sh

(See INSTALL.TXT for details on what this does/does not do.)

# Manual Installation (Debuan/Ubuntu)

- Install dependencies:     apt-get install sipcalc jq wget tshark rrdtool
- Optional arpwatch tool:   apt-get install arpwatch
- Install `./bin/ixp-watch` into `/usr/local/bin`
- Install `./bin/ixp-watch-tidy` into `/usr/local/bin`
- Install `./bin/sponge` (manual sponge tool)
-  - or `./IXP-Manager/auto_sponge` into  `/usr/local/bin` (If using IXP Manager auto sponge)
- Install `./IXP-Manager/update_ethers.sh` into `/usr/local/bin` (If using IXP Manager)
- Make sure all these scripts have execute bit set.
- Run `dpkg-reconfigure wireshark-common` and select **Yes** to the question asking if non-root users should be permitted.
- Create a user called `ixpwatch`. Add it to group `wireshark` (create this group if it doesn't exist)
- Make directories:
  - `/var/ixpwatch`
  - `/var/ixpwatch/samples`
  - `/var/ixpwatch/watch`
  - `/var/ixpwatch/www`
- (For html reports, decide output directory. Default is /var/ixpwatch/www)
- The included html_example uses server side includes, but this could be changed to
  rsync the files somewhere etc.
- Set permissions / owner so that ixpwatch can write files: `chown -R ixpwatch /var/ixpwatch`
- Set up cron jobs:
  - `*/15 * * * *        ixpwatch    /usr/local/bin/ixp-watch >/dev/null 2>&1`
  - `3 9 * * *           ixpwatch    /usr/local/bin/ixp-watch-tidy > /dev/null 2>&1`

  - If using the auto_sponge tool for automating sponges from IXP Manager:
     - `*/5 * * * *         root        /usr/local/bin/auto_sponge update`
     - `10 */3 * * *        root        /usr/local/bin/update_ethers.sh`


Changelog
* 1.17 2023-07-24 - robl
  - ixp-watch-tidy: get config options from config file.
  - ixp-watch-tidy: Option to purge reports as well as samples.
  - install.sh: prompt for purge settings.
  - Configurable report file extension. .TXT by default.

* 1.16a 2023-05-09 - robl
  - Bugfix: Do not try to copy the report via slack on arp alert, only email.
  - Bugfix: RRD Graph PNGs not updated.

* 1.16 2022-12-14 - robl
   - Move things to functions to avoid repeated code in many places.
    - Error handling / alerting is now more consistent
    - do_alert: Handle all alerts/emails/log output in function
    - log_error: Handle all error output logs in function
    - check_errors: Handle checking for log_error logs
    - die/cleanup/cleanexit: Clean up on exit, trap CTRL-C
    - required_vars: Check all variables required for proper operation are set before continuing
    - create_report/report/report_header/report_section_file - For building reports
    - do_graph: create and update rrd graphs
   - do_alert: Add slack alerter
   - Make path to logger prog config variable.
   - Configure syslog options in install script
   - Better error handling if called programs return non-zero
   - Check we can execute tshark before starting capture
   - Log stderror output from tshark capture if it fails to run correctly
   - Replace backticks with newer $(command) syntax

* 1.15.1 2021-07-06 - robl
   - Bugfix: Exit after sample file size exceeds MAX_SAMPLE_SIZE.
   - Move more options to config file
   - Make update_ethers tool use CONFIG.
   - Fix auto_sponge/sponge to use -c <config> without breaking existing cmd options
   
* 1.15 2021-06-24 - robl
   - Move example config file location and default location to avoid clash
     when installing via automation/git.
  -  Set default config location to /etc/ixpwatch
   - Make ALARM_PAGER optional.
   - [Issue #4] Move COUNTSFILE to config, reorder config vars so it works correctly.
   - Simple install script (works with debian/ubuntu to install from repository)
   - Make sponge utility use CONFIG file.
   - Make ixp-watch-tidy utility use CONFIG file.

* 1.14 - 2021-06-23 - robl
  - IPv6 is a thing now. Removed option to disable IPv6 processing.
  - Add report and alert for IPv6 Router Advertisements.
  - Fixes [Issue #3]: Use better tests -z and -n instead of == operator 
    and "" to test empty values. (This was throwing an error in some 
    environments if POSIX-incompatible shell is used)
  - Make the dead BGP sessions output more useful by resolving IPs. To use 
    old format instead, set BGPOPENS_OLD_FORMAT=1
  - Move user config variables to a seperate file to make upgrades easier.
  - Config file can now be specified with "-c <filename>" to make
    running multiple instances for multiple LANs easier.

* 1.13 - 2020-10-25 - robl
  - Run disk space checks before starting sample capture, not after. 
    This means it works if the disk is already full. It will now be run even 
    if capture fails or the script exits for some reason. It might also free 
    up enough disk space for the capture to succeed, rather than the capture
    failing and the script exiting before the disk space check/purge can run.

  - Make disk space check optional. (Don't do the check if DISK_PERCENT_PROG is 
    undefined (commented out) or empty.)

  - Bugfix: [Issue #2]: Correct non-unicast filter to use >= 224.0.0.0 and 
    not >= 223.0.0.0

* 1.12 - 2020-10-09 - robl
  - Bugfix: Improve matching on MY_IP for ICMP/sponge.

* 1.11 - 2020-08-24 - robl
  - Roll in some LONAP changes.
  - Better method of managing disk space / automatically remove samples if disk space
    becomes low.
  - Changes to ifconfig / use ip cmd instead.
  - Changes to syslog / loghost
  - Other small changes and bugfixes.
  - Update html examples.
  - Add IXP-Manager sponge automation tool and example cgi (see files in ./IXP-Manager)

* 1.10 - 2018-01-24 - robl

  - Roll in some LONAP changes.
  - Update TSHARK commands for updated version of tshark (requires explicit
    "-Y" filter flag for reading samples.)
  - Newer versions of tshark require "-f" capture filter.
  - Add IPv6 Traffic type summary to report.
  - Fixed IPv4 ICMP output (tshark output format changed)
  - Minor changes to docs and comments.
  - Included html files from LONAP as example for RRD stats page.

* 1.9 - 2012-05-17 - robl
  - I'm back. Different exchange this time...
  - Changed references to tethereal in EURO-IX released version to tshark.
  - Added basic RRDtool graph and html stats functionality.
  - Other small changes.
  - Include revised ixp-watch-tidy cron script.
  - bug fix: newer versions of tshark changed output slightly for ARP.
  - Now has GPL2 License as of 1.9 release. See LICENSE and LICENSE_OLD

* 1.8 - 2005-12-30 - robl
  - better handling of huge samples with the MAX_SAMPLE_SIZE parameter.
    Do not attempt to process a huge sample file, but just report this.
    - i.e, do not try to process a 2GB sample received if the network breaks!
  - You can optionally automatically delete huge samples to save disk space.
    - see ENORMOUS_SAMPLE.
  - run gzip with "nice -n 10" to reduce CPU on the box.
  - run GZIP process in the background and don't wait for it to finish.
  - now logs excess ARP events to syslog.
  - now logs when ARP levels return to normal to syslog.
  - exclude my own primary IP from capture samples.
  - bug fix: newer versions of tethereal changed output slightly for CDP.

* 1.7-interim 2003-10-20 - robl
  - better handling of l2tp and GRE traffic spew.
  - bug fix: sponge tool - added full path name to "ip" commands.
  - changed some of the tethereal command lines not to DNS resolve
    when reading sample files, as when a large sample file comes in, 
    this is too CPU intensive and/or takes too long to process.
  - bug fix: alarms_arp.tmp file was in the wrong location.
    (changed to $LOG_ROOT instead of $LOGDIR)
  - NB: There has been an updated release of tethereal fixes 

config.sh
#!/bin/bash

################################################################################
#### Local configuration. Things you WILL need to check                        #
################################################################################

# REPORT_EMAIL='SOMETHING@YOUR.DOMAIN.HERE' # E-mail to send regular reports to
# (or leave REPORT_EMAIL undefined if you don't want reports)

ALARM_EMAIL='YOU@YOUR.DOMAIN.HERE' # E-mail to send bad thing alarms to

ALARM_PAGER='YOU@YOUR.DOMAIN.HERE' # E-Mail to page things out

MAILPROG=/usr/bin/mail   # command to send e-mails
# (redhat this is /bin/mail by default)

GZIP='nice -n 10 /bin/gzip'       # Location of gzip prog

CAP_INTERFACE=eth0        # Interface that we run the capture on

SAMPLE_TIME=900           # Capture time in seconds (default 15 minutes)

NETWORK='YOUR_NETWORK_HERE'   # Network Name. - some name

# Local syslog facility. Leave blank to turn off.
# (facility must be set up in your syslog.conf of course.)
LOCAL_FACIL=local4.debug


# Remote syslog option. Comment out to disable it.
LOGHOST=                  # optional syslog host to send alerts to
                          # leave blank to log locally instead.
                          # Check the below "/usr/bin/logger" commands to ensure
                          # You are using "-n" or "-h" accordingly for Linux or BSD.

LOG_FACILITY=local4.debug # Facility to send logs to
                          # leave both blank to not log anywhere.

ARP_WARNLEVEL=4000        # Alert when number of ARPS Per Minute reaches this.
                          # You may not know this now, but you can use the reports
                          # to work out what is a reasonable level.

SPANNING_TREE_BAD=1       # Do you want STP frames reported? 1=Yes 0=No.

WRITE_COUNTS=0            # Write counts to /tmp/ixpwatch.$NETWORK.counts?
                          # (useful if you want to plot graphs etc) 1=Yes 0=No.
                          # (See also COUNTSFILE below - you may need to change it!)

MAX_SAMPLE_SIZE=52428800  # Maximum sensible file size to try and process (50M)

ENORMOUS_SAMPLE=`expr $MAX_SAMPLE_SIZE \* 25`  # If the file is too enormous, just delete it.

# Disk space usage for disk where samples are stored.
# (Comment this out to disable auto sample purge based on free disk space)
DISK_PERCENT_PROG='/bin/df -h --output=pcent /'
DISK_PERCENT_MAX=95   # Max disk use before deleting samples.

DO_RRD=0                  # Plot RRT?
RRDTOOL=/usr/bin/rrdtool  # If so, where is RRDtool?

# As of 1.14, we try to make the list of dead BGP sessions more helpful by resolving
# the IP address, rather than just a big list of IP addresses. If you want to use
# the old format (no resolving), set BGPOPENS_OLD_FORMAT=1.
BGPOPENS_OLD_FORMAT=0

# Directories to put things in:
# *** These need to exist for this to work properly! ***

SAMPLE_ROOT=/var/ixpwatch/samples
LOG_ROOT=/var/ixpwatch/watch
TEMP_DIR=/var/ixpwatch/tmp

# Where to put RRD files and HTML files if DO_RRD is enabled:

HTML_DIR=/var/ixpwatch/www
RRD_DIR=${HTML_DIR}/graphs
GRAPH_DIR=${HTML_DIR}/graphs

# Where to find tshark:
# (I'm going to trust this as correct.)
TSHARK=/usr/bin/tshark


IXP-Manager/update_ethers.sh
#!/bin/bash
# This file is part of IXP Watch
#
# IXP Watch is free software: you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation, version v2.0 of the License.
#
# IXP Watch is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
# for more details.
#
# You should have received a copy of the GNU General Public License v2.0
# along with IXP Watch If not, see:
#
# http://www.gnu.org/licenses/gpl-2.0.html
#
# <rob@lonap.net> - 2020-08-24
# Script to update /etc/ethers with member details to make
# IXP-Watch reports and tcpdump more useful.
# Set this up as a cron task to periodically update /etc/ethers e.g:
#
# 15 */2 * * * root /usr/local/bin/update_ethers.sh


# Where is config vars file:
CONFIG=/etc/ixpwatch/config.sh

# default ethers file (set in CONFIG if different)
ETHERSFILE=/etc/ethers

# Config file can also be specified from the command line with -c <filename>:
optstring=":c:"

while getopts ${optstring} arg; do
  case "${arg}" in
    c)
            # Set config file:
            CONFIG=${OPTARG}
    ;;
  esac
done


if [ ! -f $CONFIG ] ; then
 echo "Could not open config file $CONFIG!"
 exit 1
fi

source $CONFIG

TMPFILE=/tmp/ethers.$$

umask 002

  if [ ! -f "${ETHERSFILE}" ] ; then
    touch ${ETHERSFILE} || echo "ERROR: could not write to ${ETHERSFILE}" >&2 ; exit 1
  fi

# get updated ethers from IXP Manager:
$WGET $WGET_OPTS $URL_ETHERS | $JQ -r '.ethers[] | [.mac,.ethersdesc] | @tsv' > ${TMPFILE}

  COUNTLINES=$(grep -c "__" ${TMPFILE})

   if [ $COUNTLINES -lt 20 ] ; then
     echo "Hmm. file for ${ETHERSFILE} seems a bit short to me. Not using." >&2
     mv ${TMPFILE} ${TMPFILE}.err

   else 

    cat ${TMPFILE} > ${ETHERSFILE}

   fi

# clean up tmp file:
if [ -f "${TMPFILE}" ] ; then
	rm ${TMPFILE}
fi


install/install.sh
#!/bin/bash
# This file is part of IXP Watch
#
# IXP Watch is free software: you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by the
# Free Software Foundation, version v2.0 of the License.
#
# IXP Watch is distributed in the hope that it will be useful, but
# WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
# for more details.
#
# You should have received a copy of the GNU General Public License v2.0
# along with IXP Watch If not, see:
#
# http://www.gnu.org/licenses/gpl-2.0.html

clear

REPO_URL="https://github.com/euro-ix/IXP-Watch.git"

DEF_INSTALL_DIR="/usr/local/ixpwatch"
DEF_LINK_DIR="/usr/local/bin"
DEF_DATA_DIR="/var/ixpwatch"
DEF_CONF_DIR="/etc/ixpwatch"

DEF_PURGE_SAMPLE_DAYS=10
DEF_ZIP_REPORTS_DAYS=30
DEF_PURGE_REPORTS_DAYS=0

DEF_LOGGER="/usr/bin/logger"
DEF_LOGHOSTS="192.168.100.10 192.168.200.20"

DEF_LOCAL_FACIL="local4.debug"
DEF_LOG_FACILITY="local4.debug"

DEF_PREFIX_IPV4="5.57.80.0/22"
DEF_CAP_INTERFACE="eth1"
DEF_NETWORK="IXP-LAN"

# install these packages on debian/ubuntu by default:
DEF_PACKAGES="tshark sipcalc"

DEF_USER="ixpwatch"

RUNNING_USER=$(whoami)

echo "IXP-Watch Install Script"
echo "========================"
echo ""
echo "This script will attempt to install a working IXP-Watch with some"
echo "default settings to get started."
echo ""
echo "Please read INSTALL.TXT to understand the config options in more detail."
echo ""
echo ""

if [ $RUNNING_USER != "root" ] ; then
	echo " *** This script needs to run as root, because it needs to install packages"
	echo "     and to set the correct permissions."
	exit 1
fi

if [ -f /etc/debian_version ] ; then
 DEBIAN=$(cat /etc/debian_version)
 echo "** Debian/Ubuntu $DEBIAN found. Will ask to install apt packages."
else
 echo "** Not Debian / Ubuntu. You may need to install extra packages for tshark, rrdtool etc."
fi
echo ""

GIT=$(which git)

if [ -z "$GIT" ] ; then

    echo "*** git not found! "

	if [ -n "$DEBIAN" ] ; then
	 printf "...will install git with apt"
	 DEF_PACKAGES="git ${DEF_PACKAGES}"

	else

	 exit 1

	fi

echo ""

fi

# Get prompt to variable and make it look nice
get_prompt () {

	local PROMPT_TEXT=$1
	local PROMPT_DEFAULT=$2

	local prompt_len=${#PROMPT_DEFAULT}

	if [ $prompt_len -le 1 ] ; then
	 local OPTS="-n 1"
    else
	 local OPTS="-e -i ${PROMPT_DEFAULT}"
	fi

    local ten="          "
    local pad="$ten$ten$ten"

	if [ -n "$PROMPT_DEFAULT" ] ; then
	    local prompt="${PROMPT_TEXT}"
		prompt="${prompt:0:40}${pad:0:$((40 - ${#prompt}))} : "
        read -p "$prompt" ${OPTS}
	else
    	read -p "${PROMPT_TEXT}" ${OPTS}
	fi

	echo "${REPLY}"
}

function yes_or_no {
	local default=$2
	local defprompt="[y/n]"

	 if [ "$default" = "y" ] ; then defprompt="[Y/n]" ; local retval="1" ; fi
	 if [ "$default" = "n" ] ; then defprompt="[y/N]" ; local retval="0" ; fi

    while true; do
	    yn=$( get_prompt "$1 $defprompt" "$default" )
        case $yn in
            [Yy]*) echo "1" ; return 1 ;;  
            [Nn]*) echo "0" ; return 0 ;;
			*) echo "${retval}" ; return $retval ;;
        esac
    done
}


INSTALLDIR=$( get_prompt "Script install directory" ${DEF_INSTALL_DIR} )

echo ""
echo "Data directory setup. Must be a location with sufficient space"
echo "for example 15-20G for sample and report storage."
echo ""

DATA_DIR=$( get_prompt "Data directory"          "${DEF_DATA_DIR}" )
CONF_DIR=$( get_prompt "Config directory"        "${DEF_CONF_DIR}" )

SAMPLE_ROOT="${DATA_DIR}/samples"
LOG_ROOT="${DATA_DIR}/watch"
TEMP_DIR="${DATA_DIR}/tmp"

DEF_HTML_DIR="${DATA_DIR}/www"

PURGE_SAMPLE_DAYS=$( get_prompt "Days to keep capture sample files"   "${DEF_PURGE_SAMPLE_DAYS}" )
ZIP_REPORTS_DAYS=$( get_prompt  "Compress reports older than (days)"  "${DEF_ZIP_REPORTS_DAYS}" )

echo ""
CAP_INTERFACE=$( get_prompt "Peering LAN capture interface"  ${DEF_CAP_INTERFACE} )
PREFIX_IPV4=$( get_prompt "Peering LAN IPv4 prefix"  ${DEF_PREFIX_IPV4} )
NETWORK=$( get_prompt "Name for this peering LAN (1 word)"   ${DEF_NETWORK}       )
NETWORK=$(echo "${NETWORK}" | sed s/[^\.a-zA-Z0-9_\-]//g)

echo ""
CONFIG=$( get_prompt "Config file (will be created)"       ${CONF_DIR}/${NETWORK}.conf )
echo ""

USER=$( get_prompt "user/group to add for ixpwatch"        ${DEF_USER}   )
GROUP=${USER}

echo ""
echo "Email settings"
echo "=============="
echo ""
echo "Email for alerts notification: Address to receive IXP-Watch notifications (must be set)"
echo "Email for SMS/pager: for urgent alerts (STP and ARP storms) (optional)"
echo "You may also enable REPORT_EMAIL in the config to receive a copy of every report"
echo ""

ALARM_EMAIL=$( get_prompt "Email for alerts"  ${USER} )
ALARM_PAGER=$( get_prompt "Email for SMS/urgent alerts" ${ALARM_EMAIL} )

echo ""
echo "SYSLOG settings"
echo "==============="
echo ""
echo "IXP-Watch supports local and/or remote syslog"
echo ""

LOGGER=$( get_prompt "Location of logger" ${DEF_LOGGER} )

DO_LOCAL_SYSLOG=$( yes_or_no "Enable syslog to LOCAL host? [yes]" y) ; echo ""

if [ $DO_LOCAL_SYSLOG = 1 ] ; then

 LOCAL_FACIL=$( get_prompt "- LOCAL Syslog facility"  "${DEF_LOCAL_FACIL}" )

fi

DO_REMOTE_SYSLOG=$( yes_or_no "Enable syslog to REMOTE host(s)? [yes]" y) ; echo ""

if [ $DO_REMOTE_SYSLOG = 1 ] ; then

 LOG_FACILITY=$( get_prompt "- REMOTE Syslog facility"  "${DEF_LOG_FACILITY}" )
 LOGHOSTS=$( get_prompt "- REMOTE Syslog host(s)"       "${DEF_LOGHOSTS}" )

fi


echo ""
echo "Optional extras"
echo "==============="
DO_RRD=$( yes_or_no "Set up RRD graphs? [yes]" y) ; echo ""

if [ $DO_RRD = 1 ] ; then

 DEF_PACKAGES="${DEF_PACKAGES} rrdtool"

 HTML_DIR=$( get_prompt "- Directory for html output"       ${DEF_HTML_DIR}  )
 RRD_DIR=$( get_prompt "- Directory for RRD files"          "${DEF_HTML_DIR}/graphs" )
 GRAPH_DIR=$( get_prompt "- Directory for graph png files"  "${DEF_HTML_DIR}/graphs" )
else
 HTML_DIR=${DEF_HTML_DIR}
 RRD_DIR=${DEF_HTML_DIR}/graphs
 GRAPH_DIR=${DEF_HTML_DIR}/graphs
fi
echo ""
echo "== Creating directories:"

CREATE_DIRS="$INSTALLDIR $DATA_DIR $CONF_DIR $SAMPLE_ROOT $LOG_ROOT $TEMP_DIR $HTML_DIR $RRD_DIR $GRAPH_DIR"

for dir in ${CREATE_DIRS}

  do 

    echo "- $dir"

	if [ -d "$dir" ] ; then

	    if [ "$dir" = "$INSTALLDIR" ] ; then

		  clobber=$( yes_or_no "dir exists! replace? [yes]" y)

		  if [ "$clobber" = "1" ] ; then
		    rm -rf $dir
          else
		    echo "not replacing. Install may not work correctly!"
         fi
    echo ""
  	    fi
	fi

done

for dir in ${CREATE_DIRS}

  do 
  mkdir -p $dir || exit 1

done


if [ -n "$DEBIAN" ] ; then
  echo ""
  DO_ARPWATCH=$( yes_or_no "Install arpwatch tool? [yes]" y)
 if [ $DO_ARPWATCH = 1 ] ; then
   DEF_PACKAGES="${DEF_PACKAGES} arpwatch"
 fi
 echo ""
 echo "== Preparing apt selections..."
 echo "wireshark-common wireshark-common/install-setuid boolean true" | debconf-set-selections
 echo ""
 echo "== Installing apt packages..."
 
  for package in ${DEF_PACKAGES}
  do 
    echo "-->> installing: $package : "
    apt-get -y install $package
  done
fi

GIT=$(which git)
if [ -z "$GIT" ] ; then
 echo "** git not found! please install to continue!"
 echo ""
 exit 1
fi

echo ""
echo "== Fetching from repository into ${INSTALLDIR}"

git clone ${REPO_URL} ${INSTALLDIR} || exit 1

echo "== Create user for ixpwatch: ${USER}"
useradd -c "IXP-Watch" -d ${DEF_DATA_DIR} -s /bin/bash --no-create-home --system ${USER}
usermod -a -G wireshark ${USER}

echo ""
echo "== Setting up files:"

mkdir -p ${DEF_LINK_DIR}
for file in ${INSTALLDIR}/bin/*
do
        if [ ! -e "${DEF_LINK_DIR}/${file}" ] ; then
        ln -s $file ${DEF_LINK_DIR}
		fi
done

for file in ${INSTALLDIR}/html_example/*
do
	    if [ ! -e "${HTML_DIR}/${file}" ] ; then
		  cp ${INSTALLDIR}/html_example/* ${HTML_DIR}
		fi
done

MY_NET=$(echo "${PREFIX_IPV4}" | cut -f1 -d/)
MASK=$(echo "${PREFIX_IPV4}" | cut -f2 -d/)
SANITY_CHECK=$(echo ${MY_NET} | cut -c 1-6 )

echo "== Writing configuration to ${CONFIG}"
echo "s|^ALARM_EMAIL=.*|ALARM_EMAIL=\'${ALARM_EMAIL}\'|g" | tee /tmp/sed_cmd.$$
echo "s|^ALARM_PAGER=.*|ALARM_PAGER=\'${ALARM_PAGER}\'|g" | tee -a /tmp/sed_cmd.$$
echo "s|^NETWORK.*|NETWORK=\'${NETWORK}\'|g"              | tee -a /tmp/sed_cmd.$$
echo "s|^CAP_INTERFACE=.*|CAP_INTERFACE=${CAP_INTERFACE}|g" | tee -a /tmp/sed_cmd.$$
echo "s|^DO_RRD=.*|DO_RRD=${DO_RRD}|g" | tee -a /tmp/sed_cmd.$$
echo "s|^SAMPLE_ROOT=.*|SAMPLE_ROOT=${SAMPLE_ROOT}|g" | tee -a /tmp/sed_cmd.$$
echo "s|^LOG_ROOT=.*|LOG_ROOT=${LOG_ROOT}|g" | tee -a /tmp/sed_cmd.$$
echo "s|^TEMP_DIR=.*|TEMP_DIR=${TEMP_DIR}|g" | tee -a /tmp/sed_cmd.$$
echo "s|^HTML_DIR=.*|HTML_DIR=${HTML_DIR}|g" | tee -a /tmp/sed_cmd.$$
echo "s|^RRD_DIR=.*|RRD_DIR=${RRD_DIR}|g" | tee -a /tmp/sed_cmd.$$
echo "s|^GRAPH_DIR=.*|GRAPH_DIR=${GRAPH_DIR}|g" | tee -a /tmp/sed_cmd.$$
echo "s|^MY_NET=.*|MY_NET=${MY_NET}|g" | tee -a /tmp/sed_cmd.$$
echo "s|^MASK=.*|MASK=${MASK}|g" | tee -a /tmp/sed_cmd.$$
echo "s|^SANITY_CHECK=.*|SANITY_CHECK=${SANITY_CHECK}|g" | tee -a /tmp/sed_cmd.$$
echo "s|^LOGGER=.*|LOGGER=${LOGGER}|g" | tee -a /tmp/sed_cmd.$$

if [ $DO_LOCAL_SYSLOG = 1 ] ; then
 echo "s|^# LOCAL_FACIL=.*|LOCAL_FACIL=${LOCAL_FACIL}|g" | tee -a /tmp/sed_cmd.$$
fi

if [ $DO_REMOTE_SYSLOG = 1 ] ; then

 echo "s|^# LOGHOSTS=.*|LOGHOSTS=\'${LOGHOSTS}\'|g" | tee -a /tmp/sed_cmd.$$
 echo "s|^LOG_FACILITY=.*|LOG_FACILITY=${LOG_FACILITY}|g" | tee -a /tmp/sed_cmd.$$

fi

sed -f /tmp/sed_cmd.$$ ${INSTALLDIR}/conf/config.dist > ${CONFIG}
rm -f /tmp/sed_cmd.$$

# CONFIG var needs setting in scripts:
for file in ${INSTALLDIR}/bin/*
  do
        sed s^CONFIG=.*^CONFIG=${CONFIG}^ -i ${file}
  done

id -u ${USER} > /dev/null
USER_OK=$?
if [ "$USER_OK" = "0" ] ; then

# fix up permissions:
echo ""
echo "== Setting owner/permissions"
chown -R ${USER}:${GROUP} ${INSTALLDIR}
chown -R ${USER}:${GROUP} ${DATA_DIR}
chown -R ${USER}:${GROUP} ${CONF_DIR}
chmod a+x ${INSTALLDIR}/bin/*

	if [ -d "/etc/cron.d" ] ; then
		echo ""
		echo "== Setting up cron tasks"
		echo "*/15 * * * * ${USER} ${INSTALLDIR}/bin/ixp-watch >/dev/null 2>&1" | tee /etc/cron.d/ixp-watch
		echo "5 7 * * * ${USER} ${INSTALLDIR}/bin/ixp-watch-tidy > /dev/null 2>&1" | tee /etc/cron.d/ixp-watch-tidy

        echo "Note: sample interval is 15 minutes (900 seconds). If you change this in the IXP-Watch config,"
		echo "remember also to change the cron task /etc/cron.d/ixp-watch."

	fi

else

echo "** Warning: User ${USER} could not be created! You will need to:"
echo "  - create the user and group for ${USER}"
echo "  - fix up permissions/ownerships on the directories (see INSTALL.TXT)"
echo "  - add user to the wireshark group as required"
echo "  - install cron tasks"

fi

echo "== Setup complete!"
echo ""
echo "Further actions:"
echo "Please read INSTALL.TXT to understand the config options in more detail."
echo ""
echo "1. Review the config file ${CONFIG} and edit any further settings as needed."
echo "    - you may also need to check the ixp-watch script itself."
echo "2. If you installed arpwatch, you may want to edit the config: /etc/default/arpwatch on debian."
echo "3. Set up bin/sponge or IXP-Manager/auto_sponge tools. (optional)"
echo "4. Edit web pages in ${HTML_DIR} / set up web location."
echo ""
